### ISSUE TITLE
Configure heartbeat interval and election timeout

### ISSUE THREAD

**Issue by ChrisKujawa** [camunda/camunda#3096]
**Description**

Currently the default values for heartbeat and election timeout is quite high. Especially for the election timeout, which causes that the start up and rejoin takes a while. 

```
RaftPartitionGroupConfig {
  private static final Duration DEFAULT_ELECTION_TIMEOUT = Duration.ofMillis(2500);
  private static final Duration DEFAULT_HEARTBEAT_INTERVAL = Duration.ofMillis(250);
```

Normally the election timeout should be between 150 and 300 ms ( https://raft.github.io/raft.pdf page 6)

> Raft uses randomized election timeouts to ensure that
split votes are rare and that they are resolved quickly. To
prevent split votes in the first place, election timeouts are
chosen randomly from a fixed interval (e.g., 150–300ms).

This is probably also the reason why always the same node becomes leader - because we have no randomized election timeout.

Could look like this:
```
    final Builder partitionGroupBuilder =
        RaftPartitionGroup.builder(raftPartitionGroupName)
            .withNumPartitions(clusterCfg.getPartitionsCount())
            .withPartitionSize(clusterCfg.getReplicationFactor())
            .withMembers(getRaftGroupMembers(clusterCfg))
            .withDataDirectory(raftDirectory)
            .withElectionTimeout(Duration.ofMillis(ThreadLocalRandom.current().nextInt(150, 300)))
            .withFlushOnCommit();
```

**Comment by deepthidevaki** [comment]
The election timeout is already randomized.
```
// Set the election timeout in a semi-random fashion with the random range
    // being election timeout and 2 * election timeout.
    Duration delay = raft.getElectionTimeout().plus(Duration.ofMillis(random.nextInt((int) raft.getElectionTimeout().toMillis())));
    heartbeatTimer = raft.getThreadContext().schedule(delay, () -> {
```

**Comment by Zelldon** [comment]
Thanks for pointing this out. For reference this is done here https://github.com/zeebe-io/atomix/blob/zeebe/protocols/raft/src/main/java/io/atomix/protocols/raft/roles/FollowerRole.java#L103 .

This also explains why it sometimes takes 5 seconds to start :roll_eyes:

**Comment by npepinpe** [comment]
@deepthidevaki from my pov we have enough metrics at this point to properly evaluate new time out - please chime in if you think we're missing critical info to tune these time outs. It also seems to me that they are indeed a little too high, even for our current performance profile.

**Comment by deepthidevaki** [comment]
Yes. I think we have enough metrics. We should first expose these configs so that we can easily evaluate with different values.

**Comment by Zelldon** [comment]
It would be now possible to benchmark the timeouts, since they are exposed now.

**Comment by Zelldon** [comment]
Lets discuss this in the next planning whether we want to do that.

**Comment by npepinpe** [comment]
Leaving it in the backlog, as it's unclear what problem we want to solve here. That said, it'd be good to understand why we use such values, and it's a good performance day topic as well.

**Comment by mkavanagh** [comment]
Investigated this for a bit - I don't have a definitive conclusion but I can recap a bit of what my understanding of the current state is.

The heartbeat interval in isolation seems fine - as long as we keep it significantly (e.g. 10x) lower than election timeout we can expect plenty of heartbeats to be sent from a healthy leader within that period and be safe against e.g. transient packet loss causing unwanted elections. We don't really need it to be low in absolute terms, only relatively to the election timeout - and lowering it further without lowering election timeouts would just increase the load heartbeats generate.

For the election timeout - it's perhaps a little on the high side but well within the range of default values used in production Raft deployments, particularly where the target infrastructure is unpredictable.

Running our load tests with 1) the defaults 2) values like recommended in the original paper and 3) some aggressively low values did indicate that in well-provisioned environments we can support much lower election timeouts (200ms election timeout wasn't measurably worse than the default, and even a deliberately low 20ms timeout which triggered regular re-elections was able to make _some_ progress between them). I don't think we have a load test setup for benchmarking recovery scenarios for leader failover at the moment, but as above that could be a good topic for a performance day.

We use priority elections by default, which means some scenarios can take much longer to recover than the timeout might suggest (e.g. 5s if the preferred leader fails): this is tied in to our mechanism for distributing leaders evenly across brokers and across regions in multi-region deployments.

(@entangled90 - as I understand it you've worked a bit on improving zone awareness mechanisms. I don't think that we have changes planned which would modify the impact of priority elections on the timeouts, but perhaps you can advise otherwise).

We'll be addressing "happy-path" leadership transfers somewhat in the near future with work on coordinated leadership transfers and automatic rebalancing, but this won't reduce the leadership transfer time for e.g. restarts/shutdowns (it could provide a mechanism for us to build on, though). I haven't found much indication that time to recovery during unplanned leadership transfers (e.g. due to crashes) is a significant issue for us operationally (to @npepinpe's point about what problem we're trying to solve here), but it's hard to know if I've missed something there.

I did find a number of support cases in which we see that even the current high values can be too low to account for e.g. I/O, CPU and GC stalls in production environments - particularly where the storage infrastructure does not meet our recommendations (a current reference for this: https://camunda.github.io/zeebe-chaos/2026/06/19/Using-slow-disk-with-Camunda/). I think it will be difficult to ensure that any change to the defaults is accompanied by appropriate configuration changes in environments where latency and stalls are an issue, so we'd probably have a negative impact overall from that.

We have documented these parameters (and their default values) but we could add something to the documentation to indicate that deployments meeting our recommendations (e.g. fast SSD storage) may be able to tune the values lower for faster leadership transfers. We also don't currently seem to be overriding the defaults for cases where we do know the target infrastructure - e.g. SaaS, or IaC deployments through the [reference architectures](https://github.com/camunda/camunda-deployment-references). But the load tests by default exceed the specs of our reference architectures, so I think it's worth some more follow up testing to qualify that the election timeout could indeed be reduced for reference deployments without negative impact.

**Comment by mkavanagh** [comment]
Correction to the above: the load tests run on disks with lower throughput than the reference deployments (I confused provisioned IOPs in the latter for size-based IOPs in the former), so any results there should be conservative with relation to the reference.

I also identified a blocking issue for shipping any overrides as part of e.g. a reference deployment: #56043
